crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,811 @@
+
+        // TODO(rakita) save last finalized block inside database but for now just take
+        // tip-finalization_window
+        let (last_finalized_block_number, _) =
+            if last_canonical_hashes.len() > finalization_window as usize {
+                last_canonical_hashes[finalization_window as usize]
+            } else {
+                // it is in reverse order from tip to N
+                last_canonical_hashes.last().cloned().unwrap_or_default()
+            };
+
+        Ok(Self {
+            externals,
+            chain_id_generator: 0,
+            chains: Default::default(),
+            block_indices: BlockIndices::new(
+                last_finalized_block_number,
+                num_of_additional_canonical_block_hashes,
+                BTreeMap::from_iter(last_canonical_hashes.into_iter()),
+            ),
+            num_of_side_chain_max_size,
+            finalization_window,
+        })
+    }
+
+    /// Fork side chain or append the block if parent is the top of the chain
+    fn fork_side_chain(
+        &mut self,
+        block: SealedBlockWithSenders,
+        chain_id: ChainId,
+    ) -> Result<(), Error> {
+        let block_hashes = self.all_chain_hashes(chain_id);
+
+        // get canonical fork.
+        let canonical_fork =
+            self.canonical_fork(chain_id).ok_or(ExecError::ChainIdConsistency { chain_id })?;
+
+        // get chain that block needs to join to.
+        let parent_chain =
+            self.chains.get_mut(&chain_id).ok_or(ExecError::ChainIdConsistency { chain_id })?;
+        let chain_tip = parent_chain.tip().hash();
+
+        let canonical_block_hashes = self.block_indices.canonical_chain();
+
+        // get canonical tip
+        let (_, canonical_tip_hash) =
+            canonical_block_hashes.last_key_value().map(|(i, j)| (*i, *j)).unwrap_or_default();
+
+        let db = self.externals.sharable_db();
+        let provider = if canonical_fork.hash == canonical_tip_hash {
+            Box::new(db.latest()?) as Box<dyn StateProvider>
+        } else {
+            Box::new(db.history_by_block_number(canonical_fork.number)?) as Box<dyn StateProvider>
+        };
+
+        // append the block if it is continuing the chain.
+        if chain_tip == block.parent_hash {
+            parent_chain.append_block(
+                block,
+                block_hashes,
+                canonical_block_hashes,
+                &provider,
+                &self.externals.consensus,
+                &self.externals.executor_factory,
+            )?;
+        } else {
+            let chain = parent_chain.new_chain_fork(
+                block,
+                block_hashes,
+                canonical_block_hashes,
+                &provider,
+                &self.externals.consensus,
+                &self.externals.executor_factory,
+            )?;
+            // release the lifetime with a drop
+            drop(provider);
```

Comment:
We are taking reference of `self.db` and wrapping it around `Box<dyn StateProvider>` that contains that reference, so with drop we are releasing that reference as rust can't do it on it own so that we can call &mut fn `self.insert_block(&mut self..`.